Add Columns template options, support InnerBlock templateOptions

aduth opened the pull request:
Closes #15662
Partially addresses #15663

This pull request seeks to enhance the `InnerBlocks` component to allow developers to specify a set of template options to allow the user to select from upon inserting the block. As demonstration, this is implemented into the Columns block, allowing the user to select from a number of predefined columns layouts.

![template-options](https://user-images.githubusercontent.com/1779930/59377883-18249a80-8d21-11e9-8b6f-45775295e7c2.gif)

_**Work In Progress:**_

- "Undo" does not work as you would want it to once having selected a block template
- Needs verification about the purpose of the `InnerBlocks` `templateInProcess` state, removed here
- Needs verification about deprecation behavior for Columns block

**Implementation Notes:**

There's a question here about if or how we preserve the current default columns count (2 columns). The absence of a columns serves as a useful indicator for considering the block in a placeholder state, similar to the Image block using the absence of its `url` attribute to determine whether it is in a placeholder state. This has obvious complications for existing content which assumes the default of 2 columns. Included here is a deprecation which is intended to transition a block which assumes the default to one which defines it explicitly. There is some awkwardness here in the fact that if a user chooses to skip template selection (the "Skip" button), we still "default" to 2 columns.

Note below in "Design Notes" the consideration about customizing labels. I'd appreciate feedback about whether this is a reasonable enhancement for `InnerBlocks`, or if we should consider shifting this responsibility back to the block component to handle (optionally enhancing Placeholder instead, or creating new components to render a set of Button choices).

**Design Notes:**

There are a few notable differences from the concepts as presented by @kjellr at https://github.com/WordPress/gutenberg/issues/15663#issuecomment-496568894. For the most part, these differences come down to consequences of consistency and in considering the technical requirements of these interfaces (e.g. labels specific to "Columns" block vs. generic labels).

- There's not quite as much padding around the actual implementation as in the concept.
   - This is inherited from the `Placeholder` styling. If we want more padding, we should consider if it should be applied to all instances of `Placeholder`.
- The "Skip" link is underlined
   - This is inherited from the `<Button isLink />` styling. If we want no underline, we should again consider if it should apply to all link-styled buttons.
   - Aside: There is some prior discussion in #7534 about the role of links vs. buttons, relevant to whether this non-navigation interaction should appear as a link or as a button.
- The icon, title, and instructions of the placeholder were made to be generic, unlike the columns-specific verbiage in the concepts.
   - This is a point where I can see how it might be good to have some baseline default, but allow for customization. My worry, however, is that the technical implementation of customizing the template options is already via 3 props, and this would add 3 more (or three properties of an e.g. `templateSelectionPlaceholderProps` object prop). At that point, it begs the question whether this ought to be a common feature of `InnerBlocks` vs. custom implemented in each block vs. enhancing the `Placeholder` component to better support a set of buttons.
- The style of the icon buttons is not quite the same as proposed in the concept.
   - My choosing of the words "icon buttons" here is intention, since this seems like something we should consider to be a variant of the existing [`IconButton` component](https://github.com/WordPress/gutenberg/tree/master/packages/components/src/icon-button), at least as sharing the same semantic purpose. The ones in the concept appear quite similar to those found in the inserter menu and in the block transforms menu, and subsequent screenshots of the concept display similar buttons in the sidebar.
   - Current styling could be considered temporary, and was the minimal effort to have the buttons appear in the general shape as in the concept. I might propose that we consider whether it makes sense to create some generic component for these types of buttons, or adapt the existing `IconButton` to reuse across these multitude of use-cases.

aduth committed:
Block Editor: Add template options support to InnerBlocks

aduth committed:
Block Library: Columns: Add template options

youknowriad commented on packages/block-editor/src/components/inner-blocks/README.md:
```diff
@@ -90,6 +90,69 @@ const TEMPLATE = [ [ 'core/columns', {}, [
 
 The previous example creates an InnerBlocks area containing two columns one with an image and the other with a paragraph.
 
+### `templateOptions`
```
Honestly, I'm still uncertain about this prop. It seems like that something that could have easily been built in the columns block itself (not render `InnerBlocks` until a template is chosen).

Is the idea here to "generalize" this behavior? Is this part of the goal of the original issue? 

I can see value just trying to see if we really want to introduce it now, or maybe try to see if we need it in at least two blocks to introduce something like that.

aduth replied:
> Honestly, I'm still uncertain about this prop. It seems like that something that could have easily been built in the columns block itself (not render `InnerBlocks` until a template is chosen).
> 
> Is the idea here to "generalize" this behavior? Is this part of the goal of the original issue?

From the original comment:

>I'd appreciate feedback about whether this is a reasonable enhancement for InnerBlocks, or if we should consider shifting this responsibility back to the block component to handle (optionally enhancing Placeholder instead, or creating new components to render a set of Button choices).

From the related issue (and as evidenced by prior art in a number of plugins), there's a desire for a common behavior here in template selection. I worry about inconsistency if we start proposing that it be implemented ad hoc by each block.

As in the text above, I think there might be some way we can still put it on the block to implement, with sufficient component coverage (`Placeholder`) to _at least strongly encourage_ conforming to a consistent UI.

youknowriad replied:
I think I worry more about the growing number of props the `InnerBlocks` could and will have in the future. Maybe a first step here could be to make the `TemplatePicker` component, the sharable API, the same way we have `MediaPlaceholder`...

jorgefilipecosta replied:
Sorry, I arrived late to the party. I like the idea of providing a consistent UI for this scenario that we predict may become very common. I share the concern of increasing the number of props/responsibilities of InnerBlocks with @youknowriad.
What if we only add a new prop called placeholder to InnerBlocks (it just renders a placeholder when the template is null and no child blocks exist). Then we can expose a TemplatePlaceholder component that handles all the UI? (I think this version was exactly what @youknowriad suggested). It seems InnerBlocks should not receive a prop that is called when a template is selected and it may be good to extract that behavior. Even having a placeholder prop that receives an element rendered when the placeholder should be shown is an unnecessary prop but it is useful otherwise blocks would need to check the number of child blocks they have and conditionally render the placeholder.

kjellr commented:
🎉 This is an excellent start, @aduth! 

> There's not quite as much padding around the actual implementation as in the concept.
> This is inherited from the Placeholder styling. If we want more padding, we should consider if it should be applied to all instances of Placeholder.

Cool. We can leave as is for now. 👍 

> The "Skip" link is underlined
> Aside: There is some prior discussion in #7534 about the role of links vs. buttons, relevant to whether this non-navigation interaction should appear as a link or as a button.

Ah yes — this should've been underlined in my comp. Good call. 

Lots of good discussion in that ticket. This is essentially a "Cancel" action, which WordPress typically handles visually as a link button. Depending on the outcome of #7532, if we decide to style it differently, we should also apply a similar style to the ["Cancel" link in the image replacement flow](https://cloudup.com/cJ7gLgWu99H). 

> The icon, title, and instructions of the placeholder were made to be generic, unlike the columns-specific verbiage in the concepts.

For consistency, I do think it's important to have the Columns icon and label in there. The instructions are probably okay to be generic though. 

> Current styling could be considered temporary, and was the minimal effort to have the buttons appear in the general shape as in the concept. I might propose that we consider whether it makes sense to create some generic component for these types of buttons, or adapt the existing IconButton to reuse across these multitude of use-cases.

Yep, definitely. This feels sort of like a new component to me, but it may be possible to just build upon what we use for the block styles (and for the block inserter buttons). We're thinking through adding a better "active" state to these in https://github.com/WordPress/gutenberg/issues/15906#issuecomment-501489271, which would also come in handy here.

youknowriad commented on packages/block-library/src/columns/edit.js:
```diff
@@ -43,6 +46,75 @@ import {
+			[ 'core/column', { width: 25 } ],
+		],
+	},
+];
```
In my testing, I felt that the UI is too heavy and ideally, I'd remove one option from it. (The last one maybe)

youknowriad reviewed (commented):
This is great, I feel we're almost there. Let's try to get this into the next release.

youknowriad committed:
Remove one option from the setup state of the columns block

youknowriad commented:
@aduth is AFK this week and in an attempt to land this, I'll be making some small tweaks here.

youknowriad committed:
Mark the template options API as experimental

youknowriad commented on packages/block-editor/src/components/inner-blocks/index.js:
```diff
@@ -24,13 +24,11 @@ import DefaultBlockAppender from './default-block-appender';
 		super( ...arguments );
-		this.state = {
-			templateInProcess: !! this.props.template,
-		};
```
The e2e tests show a failure because of this. I think it ensures the InnerBlocks don't focus the placeholder to avoid creating any extra block while the template is being processed (or something like that).

@jorgefilipecosta might now more, I think it would be good to document.

Right now, I'm going to restore it.

youknowriad committed:
Fix e2e tests

youknowriad commented on packages/block-library/src/columns/edit.js:
```diff
@@ -124,15 +202,13 @@ export default withDispatch( ( dispatch, ownProps, registry ) => ( {
 		setAttributes( { columns } );
 
 		let innerBlocks = getBlocks( clientId );
-		if ( ! hasExplicitColumnWidths( innerBlocks ) ) {
```
There's a bug here in master if you try to change the number of columns from `2` to `4` (more than 1). This function only accounted for a single addition/removal.

youknowriad commented on packages/block-library/src/columns/edit.js:
```diff
@@ -43,6 +46,65 @@ import {
+		icon: <SVG width="48" height="48" viewBox="0 0 48 48" xmlns="http://www.w3.org/2000/svg"><Path fillRule="evenodd" clipRule="evenodd" d="M39 12C40.1046 12 41 12.8954 41 14V34C41 35.1046 40.1046 36 39 36H9C7.89543 36 7 35.1046 7 34V14C7 12.8954 7.89543 12 9 12H39ZM39 34V14H25V34H39ZM23 34H9V14H23V34Z" /></SVG>,
+		template: [
+			[ 'core/column' ],
+			[ 'core/column' ],
```
I removed the explicit widths from the evenly distributed templates.

mtias commented:
Love to see this!

mtias commented on packages/block-library/src/columns/edit.js:
```diff
@@ -43,6 +46,65 @@ import {
+	},
+	{
+		title: __( 'Two columns; one-third, two-thirds split' ),
+		icon: <SVG width="48" height="48" viewBox="0 0 48 48" xmlns="http://www.w3.org/2000/svg"><Path fillRule="evenodd" clipRule="evenodd" d="M39 12C40.1046 12 41 12.8954 41 14V34C41 35.1046 40.1046 36 39 36H9C7.89543 36 7 35.1046 7 34V14C7 12.8954 7.89543 12 9 12H39ZM39 34V14H20V34H39ZM18 34H9V14H18V34Z" /></SVG>,
```
Can we imagine a few more cases that go beyond layout columns here to ensure this can scale well? One that comes to mind is for https://github.com/WordPress/gutenberg/issues/15623 — imagine adding the "Post" block and choosing from "title only", "title and date", "title and excerpt", "full post" shapes.

youknowriad replied:
Makes sense, it also goes beyond the innerBlocks and could potentially be used to apply all kind of attributes.

I'm thinking we should create an issue about that describing the problem (block attributes sets) and maybe include one example in addition to the setup state of the columns block in order to come up with the best API for both use-cases. Do you have an example on mind with the current set of blocks?

Also, I can go ahead and create an issue unless you want to create it yourself to detail your thoughs?

youknowriad replied:
Also, these are other reasons to mark the APIs used as experimental for now until we cover the whole feature.

aduth replied:
> Also, these are other reasons to mark the APIs used as experimental for now until we cover the whole feature.

I think this is fair. Considering from the how this scales to other templates, and @mtias comment below at https://github.com/WordPress/gutenberg/pull/16129#discussion_r294292334, it occurs to me that this might not make sense as a prop of `InnerBlocks`, but rather more similar to the block's registration of styles (both as a setting of the block, and with its own API `registerBlockStyle`). That being said, doing so leaves open the question of how to configure options like skip links, callback function. Experimental props might be the best compromise for now (maybe in the future we still support it as a prop, but automatically provide it via context using the block's settings).

mtias commented on packages/block-library/src/columns/edit.js:
```diff
@@ -43,6 +46,65 @@ import {
+ * @constant
+ * @type {Array}
+ */
+const TEMPLATE_OPTIONS = [
```
Not for now, but this is something that should be extendable like `registerTemplate( blockName, template )`. Ideally it's modelled with full page templates in mind (either `registerBlockTemplate` and `registerTemplate` or assuming a post context instead of block name).

mtias commented on packages/block-editor/src/components/inner-blocks/template-picker.js:
```diff
@@ -0,0 +1,58 @@
+/**
```
Is this something we expect to only be used for inner blocks? Is there a use case for the selection screen to be used for pre-configured block attributes?

aduth replied:
> Is this something we expect to only be used for inner blocks? Is there a use case for the selection screen to be used for pre-configured block attributes?

That's a good point. It's unclear to me how this might look from a design perspective, since the current concepts seem to align quite closely with how _blocks_ are arranged (technically aligning to the template option), though I don't see any reason why it couldn't be that the options configure a block's own attributes. It speaks a bit to the original questions about whether this be proposed an enhancement to the Placeholder component rather than to the InnerBlocks component, which could be more easily adapted. I still think we could extract something common regardless.

At least as experimental props, we could still opt toward the Placeholder improvements if own-block attributes option selection pans out to be a good idea to pursue.

mtias commented on packages/block-editor/src/components/inner-blocks/template-picker.js:
```diff
@@ -0,0 +1,58 @@
+			instructions={ instructions }
+			className={ classes }
+		>
+			<div className="block-editor-inner-blocks__template-picker-options">
```
Should this be an `ul`?

mtias commented on packages/block-library/src/columns/edit.js:
```diff
@@ -43,6 +46,65 @@ import {
+ *
+ * @type {Number}
+ */
+const SKIPPED_TEMPLATE_DEFAULT_COLUMNS = 2;
```
Wouldn't `DEFAULT_COLUMNS` be enough?

youknowriad committed:
Rename default columns constant

youknowriad committed:
Use ul/li for a11y reasons

youknowriad reviewed (approved):
I think this is ready as a v1. A follow-up to expand this to any block attributes and to allow "registration"  of these variations will be needed.

I'd appreciate another review before moving forward.

kjellr commented:
> I think this is ready as a v1. A follow-up to expand this to any block attributes and to allow "registration" of these variations will be needed.

I think it's a decent V1 too. The one thing I think we should sort out before merge is this issue where the block empties out when you press undo: 

![test](https://user-images.githubusercontent.com/1202812/59688987-c9598380-91ac-11e9-81e2-7fbe794c541f.gif)

Since we don't yet have these layout controls in the sidebar, I could see users hitting undo to get back to that placeholder state if they'd like to re-adjust the layout.

chrisvanpatten commented:
This looks super slick. I agree w/ @mtias about the template picker being a more generic component, but that can obviously be accomplished in a follow-up.

Super nice work – excited to see this land!

youknowriad committed:
Try to improve the undo/redo behavior

youknowriad commented:
In the last commit I improved the undo/redo behavior here. It's not perfect because it requires two undo to actually undo the template selection. I think this is a bigger and separate problem regarding undos that are composed of more than one atomic change to perform the user action. 

I think we can push further by making it just a single undo level but this requires removing the "columns" attribute from the columns block and just consider the number of inner blocks as the source of truth here. This refactoring means we'd have to sunset the `has-n-columns` className. I'll plan to explore this separately but let me know if you care about this className and why.

kjellr commented:
Thanks @youknowriad — Undo is working on my end now. Clicking twice is weird, but since this generally works, it seems okay to explore separately if you want to. 

I am seeing a little regression with the latest updates though — pressing the layout buttons provides the right number of columns, but their widths always appear to be equal, regardless of the layout chosen: 

![columns](https://user-images.githubusercontent.com/1202812/59713660-ee1a1f00-91dd-11e9-9f80-b156a99564d8.gif)

youknowriad committed:
Fix block template selection

youknowriad commented:
Ok, I went ahead and removed the "columns" attribute from the columns block. It fixes the "undo" entirely and I also fixed the template selection bug.

This did grow more than I would have loved initially, I'd love some testing. Ideally we can ship this today and included in the release.

youknowriad committed:
Try to fix e2e and unit tests

youknowriad commented:
Weird here, I feel like I fixed the unit test locally but it's still failing on travis.

aduth commented on packages/block-editor/src/components/inner-blocks/template-picker.js:
```diff
@@ -0,0 +1,67 @@
+		>
+			{
+				/*
+				* Disable reason: The `list` ARIA role is redundant but
```
Not super relevant for this pull request, but is the approach we've taken that we need `role="list"` for every `ul` ? Especially considering it as a "bug" of a particular browser + screen reader combination, I wonder if this is something we could isolate to a Babel transform, to remove consideration from the individual developer. That, or an inverse ESLint rule to enforce it be applied.

Currently, it's far too easy to overlook (obvious by the fact I'd omitted it in my original proposal).

youknowriad replied:
Yeah, I know we did this for the Block Hierarchy as well. yeah, it does seem like we should have a rule for it.

aduth commented:
> Ok, I went ahead and removed the "columns" attribute from the columns block. It fixes the "undo" entirely and I also fixed the template selection bug.

What of the backwards-compatibility impact for theme developers who may have come to expect the `has-2-columns` class? Couldn't we still apply it by checking the length of `innerBlocks` at `save` time?

youknowriad commented:
> Couldn't we still apply it by checking the length of innerBlocks at save time?

I thought about it, but right now the Block API doesn't allow us to do it. (we don't receive the innerBlocks in the save function). I think this is one of those changes that could be handled using a dev note. I also asked @jasmussen about it. and there doesn't seem to be real usage for these classNames, especially with the column widths being customizable.

aduth commented on packages/block-library/src/columns/edit.js:
```diff
@@ -113,48 +209,54 @@ export default withDispatch( ( dispatch, ownProps, registry ) => ( {
+			innerBlocks = [
+				...innerBlocks,
+				...times( newColumns - previousColumns, () => {
+					return createBlock( 'core/column', );
```
```suggestion
					return createBlock( 'core/column' );
```

youknowriad committed:
Remove unnecessary comma

Co-Authored-By: Andrew Duthie <andrew@andrewduthie.com>

aduth commented on packages/e2e-tests/fixtures/blocks/core__columns__deprecated.json:
```diff
@@ -2,8 +2,8 @@
         "name": "core/columns",
-        "isValid": true,
-        "attributes": {
+		"isValid": true,
```
Guessing these were manually edited?

youknowriad replied:
Right, I can regenerate using the CLI

youknowriad committed:
regenerate fixtures

aduth commented on packages/block-library/src/columns/deprecated.js:
```diff
@@ -96,4 +101,33 @@ export default [
+		migrate( attributes, innerBlocks ) {
+			attributes = {
+				...attributes,
+				columns: innerBlocks.length,
```
If we've dropped the `columns` attribute, this should no longer apply?

aduth commented:
Aside my latest comment, and with future consideration of Dev Note, I think this looks good to land for an initial iteration. Thanks for taking it over in my absence. Given the discussion, I think there's lots of directions we may take with how "selection" applies (both as inner blocks and as a block's own attributes) or is reused (post layouts, etc). Implementing the props as experimental for this iteration should allow us some flexibility for how those iterations are explored.

aduth commented:
> It's not perfect because it requires two undo to actually undo the template selection. I think this is a bigger and separate problem regarding undos that are composed of more than one atomic change to perform the user action.

While the implementation here has since changed, the general problem is pretty well-summarized/tracked at #8119 (or, at least, a solution there should provide the necessary tools to avoid these situations for synchronous block updates as well).

youknowriad committed:
Remove the migrate function

youknowriad committed:
omit the columns attribute while migrating

youknowriad committed:
Use white backgrounds

jorgefilipecosta commented on packages/block-editor/src/components/inner-blocks/README.md:
```diff
@@ -90,6 +90,69 @@ const TEMPLATE = [ [ 'core/columns', {}, [
+- `icon` (`WPElement|string`): An element or [Dashicon](https://developer.wordpress.org/resource/dashicons/) slug to show as a visual approximation of the template.
+- `template` (`Array<Array>`): The template to apply when the option has been selected. See [`template` documentation](#template) for more information.
+
+For the template placeholder selection to be displayed, you must render `InnerBlocks` with a `template` prop value of `null`. You may track this as state of your component, updating its value when receiving the selected template via `onSelectTemplateOption`.
```
Should we use the experimental version when referencing props as past of other props documentation? (E.g: __experimentalOnSelectTemplateOption instead onSelectTemplateOption?)

aduth replied:
> Should we use the experimental version when referencing props as past of other props documentation?

Technically they shouldn't be documented at all ([per coding standards](https://github.com/WordPress/gutenberg/blob/master/docs/contributors/coding-guidelines.md#experimental-and-unstable-apis)), but I'd like to at least be able to have this writing available for future reference if we do decide to remove it.

jorgefilipecosta reviewed (commented):
I left a comment regarding having an external component for template choosing functionality.
I noticed a possible style problem when I try columns block on Twenty Fourteen theme I get list marks on each template option.
<img width="455" alt="Screenshot 2019-06-24 at 17 14 22" src="https://user-images.githubusercontent.com/11271197/60035063-e2c66800-96a3-11e9-90fa-f5b20d6ceb50.png">
Regarding the functionality, it seems to work correctly

youknowriad committed:
Clarify experimental props in the docs

youknowriad committed:
Increase the specificity of the list styling in the TemplatePicker

aduth commented on packages/e2e-tests/plugins/templates.php:
```diff
@@ -26,7 +26,7 @@ function gutenberg_test_templates_register_book_type() {
 			array(
 				'core/columns',
-				array(),
+				array( 'columns' => 2 ),
```
This shouldn't be here, should it?

youknowriad replied:
Right, it's probably ignored.

klihelp commented:
Thank you!:)
